mewarpx/mewarpx/testing_util.py

```diff
@@ -67,3 +70,120 @@ def change_to_warpxdir(wd):
+        margin (float): Multiplicative safety margin
+
+    Returns:
+        minwmargin (float): If min < 0, this is min*(1-margin). If min > 0, this
+            is min*(1+margin).
+    """
+    amin = np.min(array)
+    return amin*(1.0 - np.sign(amin)*margin)
+
+
+def get_max_wmargin(array, margin=0.1):
+    """Get a maximum value with a multiplicative margin of safety.
+
+    Arguments:
+        array (np.ndarray): Array to find maximum of
+        margin (float): Multiplicative safety margin
+
+    Returns:
+        maxwmargin (float): If max < 0, this is max*(1-margin). If max > 0, this
+            is max*(1+margin).
+    """
+    amax = np.max(array)
+    return amax*(1.0 + np.sign(amax)*margin)
+
+
+def test_df_vs_ref(testname, df, suffix=None, margin=0.1):
+    """Handle loading and test a reference DataFrame against test results.
+
+    This creates or appends to a ``<testname>.csv`` or
+    ``<testname>_<suffix>.csv`` file in ``tests/temp_files/Result_Stats``. It
+    then compares the results to a dataframe stored in
+    ``tests/test_files/Result_Stats`` under the same filename. This means that
+    running and failing the test many times will generate a csv file with all
+    the information needed to check that stats don't change in the future. See
+    e.g.  ``test_runsuite_vweights.py::test_runsuite_vweights2diag``.
+
+    Arguments:
+        testname (str): Name of the test
+        df (pandas.DataFrame): DataFrame containing the test results to compare
+            against a reference
+        suffix (str): An append to test names, allowing multiple DataFrames to
+            be referenced in a single test.
+        margin (float): A multiplicative safety margin: The tested df values
+        should be within `(1 +- margin) * [min or max](ref_val)`. Default 0.1.
+
+    Returns:
+        equal_within_margin (bool): True if equal within the margin. False if
+        reference file does not exist or data frames do not agree, meaning that
+        the test value does not fall within min and max of reference values
+        within some error margin.
+    """
+    csvname = (
+        '{}.csv'.format(testname) if suffix is None
+        else '{}_{}.csv'.format(testname, suffix)
+    )
+
+    # Record for future use if needed
+    mwxutil.mkdir_p(os.path.join(temp_dir, 'Result_Stats'))
+    record_file = os.path.join(temp_dir, 'Result_Stats', csvname)
+    if os.path.exists(record_file):
+        df.to_csv(record_file, mode='a', header=False, index=False)
+    else:
+        df.to_csv(record_file, mode='w', header=True, index=False)
+
+    # Load authoritative records
+    record_file = os.path.join(test_dir, 'Result_Stats', csvname)
+
+    if os.path.isfile(record_file):
+        refdf = pandas.read_csv(record_file)
+        return test_df_in_margin(testdf=df, refdf=refdf, margin=margin)
+
+    else:
+        print(("Record file {} does not exist. Reference data frame "
+               "not loaded.").format(record_file))
```

Comment:
Would you mind converting this and below prints to f-strings? https://realpython.com/python-f-strings/ They're so magical I wish I had known about them earlier...